### ISSUE TITLE
Add support for custom Task (i.e. ValueTask)

### ISSUE THREAD

**Issue by MihaMarkic** [nunit/nunit#2286]
Can you add support for custom Tasks, i.e. ValueTask (new in C# 7.0).

Right now looking at Assert.ThrowsAsync.

**Comment by rprouse** [comment]
ValueTasks and the like are probably going to become more widely used, so I agree that we should add this. I am going to move to the backlog and we can look at what it would take to implement.

**Comment by MihaMarkic** [comment]
Just a note, which probably you already know, it isn't just ValueTask or Task, but a custom type can be used as well.

**Comment by rprouse** [comment]
@MihaMarkic thanks, I am aware of that and that is why I said we need to figure out what it would take to implement. Currently `Assert.ThrowsAsync` takes an `AsyncTestDelegate` which has the signature,

```C#
public delegate System.Threading.Tasks.Task AsyncTestDelegate();
```

I don't know what that would look like for a custom return type and I don't know how our `AsyncInvocationRegion` will work. It might be easy, for example, we check if a method is async by seeing if the return type is `Task` **or** if the method has the generated `AsyncStateMachineAttribute` on it. I just don't have enough experience with custom tasks yet to know.

**Comment by jnm2** [comment]
@rprouse This PR fixes another problem and at the same time gets rid of AsyncInvocationRegion and implements support for test methods returning any awaitable type: https://github.com/nunit/nunit/pull/2181

The only hard part is figuring out a public API for assertion methods which take delegates. There is no way to constrain to awaitable types because it's a duck-typed pattern.

For simplicity, what's wrong with keeping it the way it is?

```c#
Assert.ThrowsAsync(async () => await MyValueTaskReturningMethod());
```

**Comment by jnm2** [comment]
I can expand on why if it's helpful.

**Comment by MihaMarkic** [comment]
@jnm2 While that works, isn't optimal as you are creating yet another state machine.
If it icn't possible adding a generic solution, than a ValueTask overload should do I guess.

**Comment by jnm2** [comment]
@MihaMarkic I know, and I am positive that it won't be a problem. The state machine overhead is already low and you're only using it once per call to `Assert.ThrowsAsync`.

If it turns out down the road that people have specific real-world examples where the testing performance is unacceptable, we can look at more API work.

**Comment by MihaMarkic** [comment]
It's also code verbosity and typing unnecessary keywords,
() => ...
vs
async () => await ...

**Comment by jnm2** [comment]
Thw keywords are necessary for compile-time type safety, unless we add every possible awaitable type as an overload that we want to support. At least until the Shapes and Extensions proposal at dotnet/csharplang gives use the ability to take any `SAwaitable`.

So we either drop compile-time safety, or we accept just any type and throw at runtime if it isn't awaitable. Or we may end up making an exception for ValueTask. I'm leery of doing that I'm particular because it means referencing the System.ValueTask package from the framework. I'm afraid the same thing will happen as is currently happening with System.ValueTuple: if you're targeting net47 and using a NuGet package that has a dependency on System.ValueTuple, you can't use C# tuples or the ValueTuple type in that project because the compiler sees duplicate definitions in multiple assemblies and gives cryptic errors.
Of course if the .NET team does their job properly, they'll have a properly-type-forwarding version in place *before* the RTM of the .NET Framework which adds the ValueTask type, unlike what they've done so far with the ValueTuple type.

So I could be convinced but it could mean a sudden need at an unknown time in the future for the overhead of a whole new net471- or net48-specific dependency group duplicating net45 but leaving out System.ValueTuple for the NuGet packagers of the @nunit/framework-team.

**Comment by jnm2** [comment]
I and others have [demonstrated](https://github.com/dotnet/roslyn/issues/20135#issuecomment-310673452) that this is a problem with ValueTuple for example if you reference BenchmarkDotNet in a net47 project. I don't want it to happen with ValueTask and I don't want it do be happening now with ValueTuple. If the csproj SDK for net47 is not able to resolve the problem, they should have synchronized shipping a type-forwarding version of the System.ValueTuple package for net47 with the initial RTW of net47. (/cc @jcouv)

**Comment by jcouv** [comment]
@jnm2 what you suggest advice (shipping package with type forwards *before* the RTW of the new framework) was just not possible. 
Type forwards need some RTM bits to point to, so the framework must ship first. Note that an updated ValueTuple package was published (with pre-release flag) very shortly after the framework was RTW. 

The alternatives I can see have their own problems: 
- don't ship any package until all the frameworks are published with the new types
- have the compiler deal with the ambiguity (we considered this is detail and concluded this has bigger downsides)

All this to say that if/when ValueTask is implemented in the desktop framework, the updated package will also be made available shortly after the framework is released.

**Comment by jnm2** [comment]
@jcouv Thanks for responding. The pre-release is the problem. NUnit Framework can't RTM with a pre-release dependency if this scenario repeats with System.ValueTask. I understand it can't actually be before the .NET Framework bits exist, but it should be no more than a day or two later that an RTM System.ValueTuple package with type forwards is available.

**Comment by jnm2** [comment]
And forcing users to work around the compilation problem by referencing a prerelease package is exactly as bad as having a direct NuGet dependency on a prerelease package, only harder to understand.

**Comment by jnm2** [comment]
I certainly don't understand why it's such a low priority. It's been two and a half months since people started installing the RTW bits.

**Comment by jcouv** [comment]
@jnm2 I see. Yes, I think we should have tried to flip the release bit sooner. 
Currently the package is planned to be marked as release when .NET Core 2.0 ships, but I will ask if we could do it now. Let me get back to you on that. Ping me in a few days if I didn't.

**Comment by jnm2** [comment]
Thanks Julien, that would be tremendously appreciated.

**Comment by jnm2** [comment]
@nunit/framework-team Back to topic, do we want NUnit.Framework to take a dependency on the System.ValueTuple package?

If we say yes, we will have to keep an eye on the .NET Framework to make sure that when it starts shipping ValueTask, we update the version of our NuGet System.ValueTuple dependency to the latest (counting on the .NET team to have it within a day or two), or people's code will stop compiling for more than a day or two.

I could go either way.

**Comment by MihaMarkic** [comment]
Perhaps there could be an extension (i.e. NuUnit.ValueTupleMess nuget package) that works when referencing ValueTuple nuget package. And "native" support for framework's ValueTuple where there is no need to reference external packages?

**Comment by jnm2** [comment]
@nunit/framework-team @MihaMarkic Dang it, I said the wrong name. I meant ValueTask above. 🙄

**Comment by jnm2** [comment]
@MihaMarkic Assuming we're talking about ValueTask now... that would mean compiling for a new platform, net471 or net48, in addition to net45 and the others we support. I also can't think of a way that a separate NuGet reference could add a `ValueTuple` overload to `Assert.ThrowsAsync` because static extension methods aren't yet a thing in C#.

**Comment by MihaMarkic** [comment]
@jnm2 Yeah, sorry. ValueTask of course. (you mention ValueTuple as well above :)).
Good point about extending static method. If Assert isn't a static class then it could be derived from "normal" Assert. But that would at least change the namespace of Assert type being used for asserting ValueTask results.
The other option is to use AssertEx or some other naming.
Not ideal either way, right.

**Comment by jnm2** [comment]
You could create your own Assert, inheriting from NUnit's and adding methods. If you have ReSharper it will complain that you're accessing a base static method via the derived name.

**Comment by MihaMarkic** [comment]
Sure, but it'd be nice if that was baked in somehow.

**Comment by jnm2** [comment]
Psst @nunit-framework team, waiting for a word from you.

**Comment by rprouse** [comment]
I would really prefer to not take a dependency on `ValueTask` if at all possible. I would prefer to drop compile time safety for `ThrowsAsync` and throw if it is not awaitable. Are there other alternatives that I missed in these many comments?

**Comment by jnm2** [comment]
@rprouse The other alternative is to keep compile time awaitable type safety and not change anything on our side:

```c#
Assert.ThrowsAsync(async () => await MyValueTaskReturningMethod());
```

I also don't feel like it's worth taking the dependency just to be able to write `Assert.ThrowsAsync(MyValueTaskReturningMethod);`.

**Comment by jnm2** [comment]
Nor do I really feel like it's worth removing the compile-type awaitable type safety enformrcement to allow the latter, either. Until someone can demonstrate that there is a good deal of pain of some kind by having to wrap in an async lambda.

**Comment by jnm2** [comment]
Yet another option, write a helper method:

```c#
Assert.ThrowsAsync(ToTaskDelegate(MyValueTaskReturningMethod()));
```

/cc MihaMarkic I could write up a proof of concept if you like.

**Comment by jnm2** [comment]
My bad, GitHub's close button looks so much like a Cancel button 😝

**Comment by jnm2** [comment]
Closure for our ValueTuple conversation; the release version of System.ValueTuple now works properly. https://github.com/dotnet/announcements/issues/25

**Comment by jnm2** [comment]
Fixed in the prerelease feed and included in 3.12 which is just around the corner!
